plugins/catalog-backend/src/database/DefaultProviderDatabase.ts

```diff
@@ -199,6 +199,20 @@ export class DefaultProviderDatabase implements ProviderDatabase {
     }
   }
 
+  async listReferenceSourceKeys(txOpaque: Transaction): Promise<string[]> {
```

Comment:
Given that we need to add a new DB call either way, perhaps we should make it `pruneOrphanedSourceKeys(ts, knownKeys)` and have that do the delete straight away?

Reply:
I can do, but I think it is important to log when a provider is being cleaned. No other methods for database's include an option for a logger so didnt want to break that approach with this change.

Reply:
Also, by having these all in a single transaction, wouldn't this roll back all deletions if a particular provider failed?